#' @export
setGeneric("blocks", function(object){
  standardGeneric("blocks")
})

#' @export
setGeneric("bumps", function(object){
  standardGeneric("bumps")
})

#' @export
setGeneric("covariableNames", function(object){
  standardGeneric("covariableNames")
})

#' @export
setGeneric("dmrCate", function(object){
  standardGeneric("dmrCate")
})

#' Exports results data.frames to csv files.
#' 
#' Exports results to csv files. If more than one variable is present, subfolders
#' with the name of the variable are created. For each variable, four files will 
#' be generated: probeResults.csv, dmrCateResults.csv, bumphunterResults.csv
#' and blockFinderResults.csv
#' 
#' @name exportResults
#' @rdname exportResults-methods
#' @aliases exportResults 
#' @export 
#' 
#' @param object \code{MethylationResults} or \code{MethylationRegionResults} 
#' @param dir Character with the path to export.
#' @param prefix Character with a prefix to be added to all file names.
#' @param vars Character vector with the names of the variables to be exported. Note: 
#' names should be that of the model. 
#' @return Files are saved into the given folder.
#' @examples
#' if (require(minfiData)){
#' set <- prepareMethylationSet(getBeta(MsetEx)[1:10,], pheno = data.frame(pData(MsetEx)))
#' methyOneVar <- DAPipeline(set, variable_names = "sex", probe_method = "ls")
#' exportResults(methyOneVar)
#' }
setGeneric("exportResults", function(object, dir = "./", prefix = NULL, 
                                     vars = modelVariables(object)){
  standardGeneric("exportResults")
})

#' @export 
setGeneric("feats", function(object){
  standardGeneric("feats")
})

#' @export 
setGeneric("featvals", function(object){
  standardGeneric("featvals")
})

#' Get all probes related to gene
#' 
#' Given a \code{MethylationResults} and a gene name returns the results of the 
#' analysis of all the probes of the gene. 
#' 
#' @name getGeneVals
#' @rdname getGeneVals-methods
#' @export
#' @param object \code{MethylationResults}
#' @param gene Character with the name of the gene
#' @return List of data.frames with the results of the analysis of the probes belonging to 
#' the gene 
#' @examples
#' if (require(minfiData)){
#' set <- prepareMethylationSet(getBeta(MsetEx)[1:10,], pheno = data.frame(pData(MsetEx)))
#' methyOneVar <- DAPipeline(set, variable_names = "sex", probe_method = "ls")
#' getGeneVals(methyOneVar, "TSPY4")
#' }
setGeneric("getGeneVals", function(object, gene){
  standardGeneric("getGeneVals")
})

#' @export 
setGeneric("getRange", function(object){
  standardGeneric("getRange")
})

#' @export 
setGeneric("getRDA", function(object){
  standardGeneric("getRDA")
})

#' @export 
setGeneric("globalPval", function(object){
  standardGeneric("globalPval")
})


#' @export 
setGeneric("globalR2", function(object){
  standardGeneric("globalR2")
})

#' @export
setGeneric("model", function(object){
  standardGeneric("model")
})

#' @export
setGeneric("modelVariables", function(object){
  standardGeneric("modelVariables")
})

#' Plot a Manhattan plot with the probe results
#' 
#' Plot log p-value for each chromosome positions. Highlighting cpgs inside a range
#' is allowed.
#' 
#' @name plotEWAS
#' @rdname plotEWAS-methods
#' @aliases plotEWAS
#' @export
#' 
#' @param object \code{AnalysisResults} or \code{AnalysisRegionResults} 
#' @param variable Character with the variable name used to obtain the probe results.
#' Note: model name should be used. Original variable name might not be valid.  
#' @param range GenomicRange whose cpgs will be highlighted
#' @param main Character with the plot title.
#' @return A plot is generated on the current graphics device.
#' @examples
#' if (require(minfiData)){
#' betas <- getBeta(MsetEx)[floor(seq(1, nrow(MsetEx), 10000)), ]
#' set <- prepareMethylationSet(betas, pheno = data.frame(pData(MsetEx)))
#' methyOneVar <- DAPipeline(set, variable_names = "sex", probe_method = "ls")
#' plotEWAS(methyOneVar)
#' }
setGeneric("plotEWAS", function(object, variable = modelVariables(object)[[1]], range = NULL, 
                                main = paste("Manhattan plot of ", variable)){
  standardGeneric("plotEWAS")
})

#' Plot RDA results
#' 
#' @name plotRDA
#' @rdname plotRDA-methods
#' @aliases plotRDA
#' @export
#' 
#' @param object \code{AnalysisRegionResults}
#' @param n_feat Numeric with the number of cpgs to be highlighted.
#' @param main Character with the plot title.
#' @return A plot is generated on the current graphics device.
#' @examples
#' if (require(minfiData) & require(GenomicRanges)){
#' set <- prepareMethylationSet(getBeta(MsetEx), pheno = data.frame(pData(MsetEx)))
#' range <- GenomicRanges::GRanges(seqnames=Rle("chrY"), 
#' ranges = IRanges(3000000, end=12300000))
#' rangeNoSNPs <- DARegionAnalysis(set, variable_names = "sex", range = range)
#' plotRDA(rangeNoSNPs)
#' }
setGeneric("plotRDA", function(object, n_feat = 5, main = "RDA plot"){
  standardGeneric("plotRDA")
})

#' Plot of the region
#' 
#' Plot of the beta values againts their position. Data is taken from probe analysis.
#' Cpgs with a p-value smaller than 0.05 (without adjusting) are blue and points
#' with a p-value greater than 0.05 are red.
#' 
#' @name plotRegion
#' @rdname plotRegion-methods
#' @aliases plotRegion
#' @export
#' 
#' @param object \code{AnalysisResults} or \code{AnalysisRegionResults} 
#' @param variable Character with the variable name used to obtain the probe results.
#' Note: model name should be used. Original variable name might not be valid.  
#' @param range GenomicRange whose cpgs will be shown (only for \code{AnalysisResults}
#' objects)
#' @param main Character with the plot title.
#' @return A plot is generated on the current graphics device.
#' @examples
#' if (require(minfiData) & require(GenomicRanges)){
#' set <- prepareMethylationSet(getBeta(MsetEx), pheno = data.frame(pData(MsetEx)))
#' range <- GenomicRanges::GRanges(seqnames=Rle("chrY"), 
#' ranges = IRanges(3000000, end=12300000))
#' rangeNoSNPs <- DARegionAnalysis(set, variable_names = "sex", range = range)
#' plotRegion(rangeNoSNPs)
#' }
setGeneric("plotRegion", function(object, variable = modelVariables(object)[[1]], range = NULL,
                                  main = paste("Region plot of ", variable)){
  standardGeneric("plotRegion")
})

#' Plot R2 region values
#' 
#' @name plotRegionR2
#' @rdname plotRegionR2-methods
#' @aliases plotRegionR2
#' @export
#' 
#' @param object \code{MethylationRegionResults}
#' @param feat Numeric with the index of the  feature or character with its name.
#' @param ... Further arguments passed to plotLM
#' @return A plot is generated on the current graphics device.
setGeneric("plotRegionR2", function(object, feat, ...){
  standardGeneric("plotRegionR2")
})

#' QQ plot of probe analysis
#' 
#' Generate a QQ plot using probe results.
#' @name plotQQ
#' @rdname plotQQ-methods
#' @aliases plotQQ 
#' @export
#' 
#' @param object \code{AnalysisResults} or \code{AnalysisRegionResults} 
#' @param variable Character with the variable name used to obtain the probe results.
#' Note: model name should be used. Original variable name might not be valid.
#' @param main Character with the plot title.
#' @return A plot is generated on the current graphics device.
#' @examples
#' if (require(minfiData)){
#' betas <- getBeta(MsetEx)[floor(seq(1, nrow(MsetEx), 10000)), ]
#' set <- prepareMethylationSet(betas, pheno = data.frame(pData(MsetEx)))
#' methyOneVar <- DAPipeline(set, variable_names = "sex", probe_method = "ls")
#' plotQQ(methyOneVar)
#' }
setGeneric("plotQQ", function(object, variable = modelVariables(object)[[1]], 
                              main = paste("QQplot of", variable, "analysis")){
  standardGeneric("plotQQ")
})

#' Make a Volcano plot with the probe results
#' 
#' Plot log p-value versus the change in expression/methylation.
#' 
#' @name plotVolcano
#' @rdname plotVolcano-methods
#' @aliases plotVolcano
#' @export
#' 
#' @param object \code{MethylationResults} or \code{MethylationRegionResults} 
#' @param variable Character with the variable name used to obtain the probe results.
#' Note: model name should be used. Original variable name might not be valid.  
#' @param mindiff Numeric with the minimum change in methylation or expression needed to be significant
#' @param main Character with the plot title.
#' @return A plot is generated on the current graphics device.
#' @examples
#' if (require(minfiData)){
#' betas <- getBeta(MsetEx)[floor(seq(1, nrow(MsetEx), 10000)), ]
#' set <- prepareMethylationSet(betas, pheno = data.frame(pData(MsetEx)))
#' methyOneVar <- DAPipeline(set, variable_names = "sex", probe_method = "ls")
#' plotVolcano(methyOneVar)
#' }
setGeneric("plotVolcano", function(object, variable = modelVariables(object)[1], mindiff = NULL,
                                   main = paste("Volcano plot of", variable, "results")){
  standardGeneric("plotVolcano")
})

#' @export
setGeneric("probeResults", function(object, drop = TRUE){
  standardGeneric("probeResults")
})

#' @export 
setGeneric("RDAPval", function(object){
  standardGeneric("RDAPval")
})

#' @export 
setGeneric("regionR2", function(object){
  standardGeneric("regionR2")
})

#' @export
setGeneric("regionResults", function(object){
  standardGeneric("regionResults")
})

#' Get the top features associated with the RDA
#' 
#' Get a list of the features significantly associated to the first two RDA components
#' 
#' @name topRDAhits
#' @rdname topRDAhits-methods
#' @aliases topRDAhits
#' @export
#' 
#' @param object \code{AnalysisRegionResults} 
#' @param pval numeric with the p-value threshold. Only features with a p-values below this threshold will be shown. 
#' @return data.frame with the features, the component, the correlation and the p-value
#' @examples
#' if (require(minfiData) & require(GenomicRanges)){
#' set <- prepareMethylationSet(getBeta(MsetEx), pheno = data.frame(pData(MsetEx)))
#' range <- GenomicRanges::GRanges(seqnames=Rle("chrY"), 
#' ranges = IRanges(3000000, end=12300000))
#' rangeNoSNPs <- DARegionAnalysis(set, variable_names = "sex", range = range)
#' topRDAhits(rangeNoSNPs)
#' }
setGeneric("topRDAhits", function(object, pval = 0.05){
  standardGeneric("topRDAhits")
})


#' @export
setGeneric("variableNames", function(object){
  standardGeneric("variableNames")
})